src/profile-logic/symbol-store.js

```diff
@@ -4,139 +4,231 @@
+    // On the other hand, if we fire off a separate request for each library,
+    // latency also suffers because of per-request overhead and pipelining limits.
+    const chunks = _splitIntoChunksOfMaxValue(
+      requestsForNonCachedLibs,
+      10000,
+      ({ addresses }) => addresses.size
+    );
+
+    // Kick off the requests to the symbolication API, and create a flattened
+    // list of promises, one promise per library. Even for libraries that are
+    // handled within the same request to the symbolication API, each library's
+    // promise can fail independently if the symbol server does not have symbols
+    // for this library.
+    const libraryPromises = chunks.reduce(
+      (accum, requests) =>
+        accum.concat(
+          this._symbolProvider
+            .requestSymbolsForAddresses(requests)
+            .map((resultsPromise, i) => ({
+              request: requests[i],
+              resultsPromise,
+            }))
+        ),
+      []
+    );
+
+    // Now that the requests to the server have been kicked off, process
+    // symbolication for the libraries for which we found symbol tables in the
+    // database. This is delayed until after the request has been kicked off
+    // because it can take some time.
+    for (const { request, symbolTable } of requestsForCachedLibs) {
+      successCb(
+        request,
+        this._readSymbolsFromSymbolTable(request.addresses, symbolTable)
+      );
+    }
+
+    // Process the results from the symbolication API request, as they arrive.
+    // For each library that was not successfully symbolicated, fall back to
+    // requesting a whole symbol table from the add-on. The add-on will attempt
+    // to dump symbols from the binary.
+    await Promise.all(
+      libraryPromises.map(({ request, resultsPromise }) =>
+        resultsPromise.then(
+          results => successCb(request, results), // Success!
+          () => {
+            // The symbolication API did not have any symbols for this library.
+            // Request a symbol table from the symbol provider. The result of
+            // this request gets stored in the database.
+            const { lib, addresses } = request;
+            return this._symbolProvider.requestSymbolTable(lib).then(
+              symbolTable => {
+                // The symbol provider was able to generate a symbol table for
+                // this library. Use it to look up the addresses.
+                const results = this._readSymbolsFromSymbolTable(
+                  addresses,
+                  symbolTable
+                );
+                successCb(request, results); // Success!
+
+                return this._storeSymbolTableInDB(lib, symbolTable);
+              },
+              error => {
+                // None of the symbolication methods were successful.
+                // Call the error callback.
+                errorCb(request, error);
```

Comment:
I'm not sure. I've left it the way it was. Once we have a plan for what to do with these errors, we can reconsider this, because we'll have an expectation from an API consumer's perspective.